### DISCUSSION TITLE
Enclosure control range by mouse

### DISCUSSION THREAD

**Discussion by eturpault** [GrandOrgue/grandorgue#1381] [General]
Currently in GrandOrgue the control of the enclosure position by the mouse is surprising and not natural according to me : 
- on the bottom third part of the enclosure graphical area, the mouse vertical dragging action permits to control the full range of the enclosure in the same direction as the mouse move.
- on the middle third part of the enclosure graphical area, the mouse vertical dragging action has no effect.
- on the top third part of the enclosure graphical area, the mouse vertical dragging action permits to control the full range of the enclosure in the opposite direction as the mouse move.

Why is it implemented in this way ?
Shouldn't it be more natural to have the enclosure position controlled progressively from min to max position by the mouse vertical dragging action from the bottom to top of the graphical area ?
With Hauptwerk sample set converted to GO ODF format (will be available from OdfEdit v2.2), the sliders permitting to control some volumes (noises, blower, surround channels) are difficult to control due to these surprising controlling ranges. It is difficult as well to control the normal enclosure pedals due to the short vertical range available (a third of the full height) to make this action.

**Comment by eturpault** [comment]
Using the attributes MouseAxisStart and MouseAxisEnd of the Enclosure doen't permit to improve the action range of the mouse on the enclosure graphical area, it is rather worse.

**Reply by larspalo** [reply]
I guess you're aware that you can also use mouse (wheel or two finger touch) scrolling to smoothly change the enclosure pedal positioning? I greatly prefer that to dragging with the mouse, which I feel is very imprecise compared to the scrolling. As to the question why it's implemented that way I have unfortunately no idea.

**Reply by eturpault** [reply]
I was not aware about this possibility with the mouse wheel, it is indeed precise and it works well, thanks for the info !

**Comment by eturpault** [comment]
Maybe I have to open an issue for that, when the attribute AmpMinimumLevel of an enclosure it set at 0 and a sound is continuously playing (i.e. note pressed, blower enabled), we move this enclosure at its minimum position then we increase the enclosure position, the sound stays muted. It is necessary to stop then restart the sound to hear it again. It is systematical. If AmpMinimumLevel is set at 1, there is no issue.

**Reply by larspalo** [reply]
@eturpault That is actually another issue, but it's by design. When the amplitude of a sample that's playing reaches 0 it's interpreted as if the sound is stopped and the assigned sampler is "released" back to the pool. At least until that behavior is changed one should never use an amplitudeminlevel of 0.

**Reply by oleg68** [reply]
> but it's by design. When the amplitude of a sample that's playing reaches 0 it's interpreted as if the sound is stopped and the assigned sampler is "released" back to the pool. At least until that behavior is changed one should never use an amplitudeminlevel of 0.

The design looks like odd. It would be better to take the amplitude of the source sample, not the resulting amplitude after the enclosure.

**Reply by larspalo** [reply]
It's in the sound engine. If I remember correctly, it doesn't differentiate between when a sample amplitude reaches 0 or when a decaying release reaches 0 and subsequently is released back to the pool. That's the reason to the effect of that even if you hold down the keys of the keyboard it will be mute and stay muted until re-struck when an enclosure forces the amplitude down to zero.

**Reply by eturpault** [reply]
Thanks @larspalo for the explanations. So for converting HW to GO ODF enclosures and volume sliders OdfEdit v2.2 will put AmpMinimumLevel at 1 instead of 0 in order to avoid this sound stop effect. Anyway at enclosure minimum position the sound is attenuated enough to not be heared. Here is an example of how volume sliders render in GrandOrgue for the Sonus Paradisi sample set St. Martini Groningen in the Mixer panel :
![sliders level](https://user-images.githubusercontent.com/51839171/218250659-5480f38d-0080-45a5-ab78-29598f4b45a5.png)

**Reply by eturpault** [reply]
Another specificity of the enclosure in GO has to be taken into account when converting HW to GO ODF, to avoid an "Out of range value" error on sample set loading, due to enclosure mouse rectangle vertical dimension. In the GO src file GOGUIEnclosure.cpp, the attribute MouseRectTop is set at 13 by default (instead of 0 for Button object) and the MouseRectHeight maximum allowed value is Height - MouseRectTop - 3 (instead of Height - MouseRectTop for Button object), whereas in HW ODF the mouse rectangle can go from 0 to the height of the bitmap vertical dimension of a continuous control. So I implemented in OdfEdit a workaround to avoid this error.

**Reply by eturpault** [reply]
In the GrandOrgue help, in Enclosure objects chapter, it is written that the attribute BitmapCount has an allowed range from 1 to 127. Looking at the source code (GOGUIEnclosure.cpp) the maximum value is in reallity 128, which is better since HW sample sets use 128 steps for their volume sliders. Could the help be updated accordingly ? Thanks.
